modules/core/src/main/java/org/apache/ignite/internal/processors/cache/ClusterCachesInfo.java

```diff
@@ -1616,37 +1616,35 @@ else if (!GridFunc.eqNotOrdered(desc.schema().entities(), locQryEntities))
                 cfg.getNearConfiguration() != null);
         }
 
-        updateRegisteredCachesIfNeeded(patchesToApply, cachesToSave, hasSchemaPatchConflict);
+        if (!hasSchemaPatchConflict)
+            updateRegisteredCaches(patchesToApply, cachesToSave);
     }
 
     /**
-     * Merging config or resaving it if it needed.
+     * Merging config, resaving it if it needed.
      *
      * @param patchesToApply Patches which need to apply.
      * @param cachesToSave Caches which need to resave.
-     * @param hasSchemaPatchConflict {@code true} if we have conflict during making patch.
-     */
-    private void updateRegisteredCachesIfNeeded(Map<DynamicCacheDescriptor, QuerySchemaPatch> patchesToApply,
-        Collection<DynamicCacheDescriptor> cachesToSave, boolean hasSchemaPatchConflict) {
-        //Skip merge of config if least one conflict was found.
-        if (!hasSchemaPatchConflict) {
-            boolean isClusterActive = ctx.state().clusterState().active();
-
-            //Merge of config for cluster only for inactive grid.
-            if (!isClusterActive && !patchesToApply.isEmpty()) {
-                for (Map.Entry<DynamicCacheDescriptor, QuerySchemaPatch> entry : patchesToApply.entrySet()) {
-                    if (entry.getKey().applySchemaPatch(entry.getValue()))
-                        saveCacheConfiguration(entry.getKey());
-                }
+     */
+    private void updateRegisteredCaches(
+        Map<DynamicCacheDescriptor, QuerySchemaPatch> patchesToApply,
+        Collection<DynamicCacheDescriptor> cachesToSave
+    ) {
+        // Store config only if cluster is nactive.
+        boolean isClusterActive = ctx.state().clusterState().active();
 
-                for (DynamicCacheDescriptor descriptor : cachesToSave)
-                    saveCacheConfiguration(descriptor);
-            }
-            else if (patchesToApply.isEmpty()) {
-                for (DynamicCacheDescriptor descriptor : cachesToSave)
-                    saveCacheConfiguration(descriptor);
+        //Merge of config for cluster only for inactive grid.
+        if (!patchesToApply.isEmpty()) {
+            for (Map.Entry<DynamicCacheDescriptor, QuerySchemaPatch> entry : patchesToApply.entrySet()) {
+                if (entry.getKey().applySchemaPatch(entry.getValue()) && !isClusterActive)
+                    saveCacheConfiguration(entry.getKey());
```

Comment:
Method `saveCacheConfiguration` also called from `processJoiningNode` without state check. Do we need to add this check to `processJoiningNode` too?
Why do we need state check at all? If it was stored on inactive cluster state in 2/3 of cases, perhaps, it should be stored in all cases? Does cache configuration stored after activation? For example, new node has joined the inactive persistent cluster. Cluster has caches with patches to apply. After activation cluster works some time and shutted down. Will new query entities be available on the new node if it will be started as first node?

Reply:
This is a good question about storing only on active cluster. I'll try to store every time.